File: AF/colocviu/t2_lab/kruskal.py
from typing import List

with open("apm.in", "r") as f:
    n, m = map(int, f.readline().split())
    graf = []

    for i in range(m):
        x, y, c = map(int, f.readline().split())
        graf.append([x, y, c, 0])

    graf.sort(key=lambda x: x[2])

h = [0 for _ in range(n + 1)]
tata = [0 for _ in range(n + 1)]


def Union(x: int, y: int):
    rx, ry = FindRepr(x), FindRepr(y)
    if h[rx] > h[ry]:
        tata[ry] = rx
    else:
        tata[rx] = ry
        if h[rx] == h[ry]:
            h[ry] += 1


def FindRepr(x: int) -> int:
    if tata[x] == 0:
        return x
    tata[x] = FindRepr(tata[x])
    return tata[x]


s = 0
nrmuchii = 0
for i in range(m):
    if nrmuchii == n - 1:
        break
    p = FindRepr(graf[i][0])
    q = FindRepr(graf[i][1])

    if p != q:
        s += graf[i][2]
        graf[i][3] = 1
        nrmuchii += 1
        Union(p, q)

print(tata, h, sep="\n")

with open("apm.out", "w") as f:
    f.write(f"{s}\n{n - 1}\n")
    for i in range(m):
        if graf[i][3] == 1:
            f.write(f"{graf[i][0]} {graf[i][1]}\n")

File: AF/CRACIUN - Florin/kruskal.cpp
/*


#include <iostream>
#include <queue>
#include <algorithm>
using namespace std;

int n, m, k, B, h[100001] = { 0 }, tata[100001];

struct Muchie {
    int x;
    int y;
    long long k;
    bool viz;
};
vector<Muchie> graf;
vector<Muchie> solutie;

bool sortMuchie(Muchie& a, Muchie& b) {
    if (a.k != b.k) return a.k < b.k;
    else {
        return a.x < b.x;
    }
}

void printMuchie(Muchie& x) {
    cout << x.x << " " << x.y << " " << x.k << " vizitata: " << x.viz << endl;
}

int FindRepr(int x) {
    if (tata[x] == -1) {
        return x;
    }
    tata[x] = FindRepr(tata[x]);
    return tata[x];
}

void Union(int x, int y) {
    int rx = FindRepr(x), ry = FindRepr(y);
    if (h[rx] > h[ry]) {
        tata[ry] = rx;
    } else {
        tata[rx] = ry;
        if (h[rx] == h[ry]) {
            ++h[ry];
        }
    }
}


int main() {
    int a, b, c, nrmuchii = 0, p, q;
    long long sol = 0;
    cin >> n >> m >> k >> B;
    if (k > n) {
        cout << "Nu";
        return 0;
    }

    for (int i = 0; i <= n; ++i) {
        tata[i] = -1;
    }

    Muchie temp;
    for (int i = 0; i < m; ++i) {
        cin >> a >> b >> c;
        temp.x = a;
        temp.y = b;
        temp.k = c;
        temp.viz = false;
        graf.push_back(temp);
    }

    sort(graf.begin(), graf.end(), sortMuchie);

    for (int i = 0; i < k; ++i) {
        cin >> a >> b;
        for (auto& uv : graf) {
            if (uv.x == a && uv.y == b) {
                uv.viz = true;
                ++nrmuchii;
                sol += uv.k;

                if (sol > B) {
                    cout << "Nu";
                    return 0;
                }

                p = FindRepr(uv.x);
                q = FindRepr(uv.y);
                if (p != q) Union(p, q);
                solutie.push_back(uv);
            }
        }
    }


//    for (auto x : graf) {
//        printMuchie(x);
//    }

    for (auto& uv : graf) {
        if (nrmuchii == n - 1)
            break;
        p = FindRepr(uv.x);
        q = FindRepr(uv.y);

        if (p != q && !uv.viz) {
            sol += uv.k;
            uv.viz = true;
            ++nrmuchii;
            Union(p, q);
            solutie.push_back(uv);
        }

        if (sol > B) {
            cout << "Nu";
            return 0;
        }

    }

    cout << "Da\n";
    for (auto x : solutie) {
        cout << x.x << " " << x.y << endl;
    }

    return 0;
}*/